### PR TITLE
#6548 2PC recovery is extremely ineffective on a cluster with multiple DATABASEs fix

### LINKED ISSUE #6548: 2PC recovery is extremely ineffective on a cluster with multiple DATABASEs
When there is a crash on a cluster with multiple `DATABASE`s, the prepared transactions recovery takes a lot of time or even gets stuck, blocking the new transactions. It is kind of tedious to describe the steps, so [here](https://github.com/ivyazmitinov/citus-prepared-transactions-recovery-bug) is a reproducible example based on the [multi_transaction_recovery.sql](https://github.com/citusdata/citus/blob/main/src/test/regress/sql/multi_transaction_recovery.sql), and here are the logs after supposed crash that highlight the issue:

```
2022-12-07T09:45:36.156342257Z [ LOG:  recovered a prepared transaction on citus-prepared-transactions-recovery-bug-worker:5432
2022-12-07T09:45:36.156389244Z [ CONTEXT:  ROLLBACK PREPARED 'citus_0_db8_99_should_abort'
2022-12-07T09:45:36.156393344Z 	Citus maintenance daemon for database 22349 user 10
2022-12-07T09:45:36.156442907Z [ WARNING:  prepared transaction belongs to another database
2022-12-07T09:45:36.156450521Z [ HINT:  Connect to the database where the transaction was prepared to finish it.
2022-12-07T09:45:36.156454955Z [ CONTEXT:  while executing command on citus-prepared-transactions-recovery-bug-worker:5432
2022-12-07T09:45:36.156459130Z 	Citus maintenance daemon for database 22349 user 10
2022-12-07T09:45:36.156483892Z [ LOG:  maintenance daemon recovered 1 distributed transactions
2022-12-07T09:45:36.156493193Z [ CONTEXT:  Citus maintenance daemon for database 22349 user 10
2022-12-07T09:45:36.158067973Z [ WARNING:  prepared transaction belongs to another database
2022-12-07T09:45:36.158080817Z [ HINT:  Connect to the database where the transaction was prepared to finish it.
2022-12-07T09:45:36.158086304Z [ CONTEXT:  while executing command on citus-prepared-transactions-recovery-bug-worker:5432
2022-12-07T09:45:36.158091149Z 	Citus maintenance daemon for database 20129 user 10
2022-12-07T09:45:36.158113190Z [ WARNING:  prepared transaction belongs to another database
2022-12-07T09:45:36.158123265Z [ HINT:  Connect to the database where the transaction was prepared to finish it.
2022-12-07T09:45:36.158127393Z [ CONTEXT:  while executing command on citus-prepared-transactions-recovery-bug-worker:5432
2022-12-07T09:45:36.158130581Z 	Citus maintenance daemon for database 21609 user 10
2022-12-07T09:45:36.170147789Z [ WARNING:  prepared transaction belongs to another database
2022-12-07T09:45:36.170169927Z [ HINT:  Connect to the database where the transaction was prepared to finish it.
2022-12-07T09:45:36.170177022Z [ CONTEXT:  while executing command on citus-prepared-transactions-recovery-bug-worker:5432
2022-12-07T09:45:36.170182855Z 	Citus maintenance daemon for database 19389 user 10
2022-12-07T09:45:36.170345173Z [ WARNING:  prepared transaction with identifier "citus_0_db1_4_should_abort" is busy
2022-12-07T09:45:36.170357085Z [ CONTEXT:  while executing command on citus-prepared-transactions-recovery-bug-worker:5432
2022-12-07T09:45:36.170363419Z 	Citus maintenance daemon for database 17909 user 10
2022-12-07T09:45:36.171484834Z [ LOG:  recovered a prepared transaction on citus-prepared-transactions-recovery-bug-worker:5432
2022-12-07T09:45:36.171499222Z [ CONTEXT:  ROLLBACK PREPARED 'citus_0_db1_4_should_abort'
2022-12-07T09:45:36.171505073Z 	Citus maintenance daemon for database 17169 user 10
2022-12-07T09:45:36.171509790Z [ WARNING:  prepared transaction with identifier "citus_0_db1_4_should_abort" does not exist
2022-12-07T09:45:36.171514522Z [ CONTEXT:  while executing command on citus-prepared-transactions-recovery-bug-worker:5432
2022-12-07T09:45:36.171519297Z 	Citus maintenance daemon for database 18649 user 10
2022-12-07T09:45:36.171537917Z [ WARNING:  prepared transaction belongs to another database
2022-12-07T09:45:36.171547062Z [ HINT:  Connect to the database where the transaction was prepared to finish it.
2022-12-07T09:45:36.171550724Z [ CONTEXT:  while executing command on citus-prepared-transactions-recovery-bug-worker:5432
2022-12-07T09:45:36.171553713Z 	Citus maintenance daemon for database 17169 user 10
2022-12-07T09:45:36.171566769Z [ LOG:  maintenance daemon recovered 1 distributed transactions
2022-12-07T09:45:36.171570657Z [ CONTEXT:  Citus maintenance daemon for database 17169 user 10
2022-12-07T09:45:36.176410024Z [ WARNING:  prepared transaction belongs to another database
2022-12-07T09:45:36.176419665Z [ HINT:  Connect to the database where the transaction was prepared to finish it.
2022-12-07T09:45:36.176423164Z [ CONTEXT:  while executing command on citus-prepared-transactions-recovery-bug-worker:5432
2022-12-07T09:45:36.176426485Z 	Citus maintenance daemon for database 20869 user 10
2022-12-07T09:45:36.181883269Z [ WARNING:  prepared transaction belongs to another database
2022-12-07T09:45:36.181893312Z [ HINT:  Connect to the database where the transaction was prepared to finish it.
2022-12-07T09:45:36.181924206Z [ CONTEXT:  while executing command on citus-prepared-transactions-recovery-bug-worker:5432
2022-12-07T09:45:36.181928640Z 	Citus maintenance daemon for database 23089 user 10
2022-12-07T09:45:36.188485387Z [ WARNING:  prepared transaction belongs to another database
2022-12-07T09:45:36.188498559Z [ HINT:  Connect to the database where the transaction was prepared to finish it.
2022-12-07T09:45:36.188504706Z [ CONTEXT:  while executing command on citus-prepared-transactions-recovery-bug-worker:5432
2022-12-07T09:45:36.188508460Z 	Citus maintenance daemon for database 23829 user 10

```

If my investigation is correct, the issue is that the maintenance daemon tries to process prepared transactions that belong to another database and halts on the fail, until the next `citus.recover_2pc_interval`.

A workaround would be to shut down the coordinator (to prevent new transactions) and `ROLLBACK` the transactions manually, e.g.

```sql
DO
$do$
    DECLARE
        r RECORD;
    BEGIN
        FOR r IN SELECT database, gid FROM pg_prepared_xacts
            LOOP
                PERFORM dblink_exec(format('dbname=%s', r.database), format($cmd$ROLLBACK PREPARED '%s'$cmd$, r.gid));
            END LOOP;
    END;
$do$;
```

### PR DESCRIPTION
**Pull request by ivyazmitinov**
Fixes the transaction recovery procedure by requesting prepared transactions belonging to the maintenance daemon's database.
Also, fixes `multi_transaction_recovery` test that was not executing properly when invoked via `./citus_tests/run_test.py multi_transaction_recovery`

fixes #6548

### CODE REVIEW

**Comment by ivyazmitinov** [comment]
@microsoft-github-policy-service agree

File: src/test/regress/expected/multi_transaction_recovery_multiple_databases.out [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+              2
+(1 row)
+
+SELECT current_database();
+ current_database
+---------------------------------------------------------------------
+ db2
+(1 row)
+
+SELECT pg_sleep_for('5 SECONDS');
```
Comment (marcocitus) [inline src/test/regress/expected/multi_transaction_recovery_multiple_databases.out:106]: I assume this is to wait for the maintenance daemon to start. In other test files (multi_extension.sql) we used the following:

```
CREATE OR REPLACE FUNCTION test.maintenance_worker()
    RETURNS pg_stat_activity
    LANGUAGE plpgsql
AS $$
DECLARE
   activity record;
BEGIN
    DO 'BEGIN END'; -- Force maintenance daemon to start
    -- we don't want to wait forever; loop will exit after 20 seconds
    FOR i IN 1 .. 200 LOOP
        PERFORM pg_stat_clear_snapshot();
        SELECT * INTO activity FROM pg_stat_activity
        WHERE application_name = 'Citus Maintenance Daemon' AND datname = current_database();
        IF activity.pid IS NOT NULL THEN
            RETURN activity;   
        ELSE
            PERFORM pg_sleep(0.1);
        END IF ;
    END LOOP;
    -- fail if we reach the end of this loop
    raise 'Waited too long for maintenance daemon to start';
END; 
$$;
```
would be nice to do something similar here to avoid unnecessarily slow tests

File: src/test/regress/expected/multi_transaction_recovery_multiple_databases.out [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+BEGIN;
+CREATE TABLE should_commit
+(
+    value int
+);
+SELECT 'citus_0_1234_1_0_' || :'db1_oid' AS transaction_2_worker_1_db_1_name
+\gset
+PREPARE TRANSACTION :'transaction_2_worker_1_db_1_name';
+BEGIN;
+CREATE TABLE should_be_sorted_into_middle
```
Comment (marcocitus) [inline src/test/regress/expected/multi_transaction_recovery_multiple_databases.out:157]: the original test file could have used way more comments (my bad), but the name "should_be_sorted_into_middle" was mainly relevant as a prepared transaction name to cover edge cases around binary search in the transaction recovery logic (between abort and commit)

the table name was always irrelevant and in this context it seems quite confusing since the prepared transaction name below does not sort into the middle of the list

File: src/backend/distributed/transaction/transaction_recovery.c [hunk 3]
```diff
@@ -404,7 +404,7 @@ PendingWorkerTransactionList(MultiConnection *connection)
 	int32 coordinatorId = GetLocalGroupId();
 
 	appendStringInfo(command, "SELECT gid FROM pg_prepared_xacts "
-							  "WHERE gid LIKE 'citus\\_%d\\_%%'",
+							  "WHERE gid LIKE 'citus\\_%d\\_%%' and database = current_database()",
```
Comment (marcocitus) [inline src/backend/distributed/transaction/transaction_recovery.c:407]: thanks, this makes a lot of sense. (and slightly embarrassing that we did not do this before)

File: src/test/regress/expected/multi_transaction_recovery_multiple_databases.out (hunk 1 above)
Comment (ivyazmitinov) [reply src/test/regress/expected/multi_transaction_recovery_multiple_databases.out]: Done

File: src/test/regress/expected/multi_transaction_recovery_multiple_databases.out (hunk 2 above)
Comment (ivyazmitinov) [reply src/test/regress/expected/multi_transaction_recovery_multiple_databases.out]: Addressed: removed those as not relevant to the test case

**Comment by marcocitus** [comment]
Would you mind fixing this test failure?
```
diff -dU10 -w /home/circleci/project/src/test/regress/expected/multi_transaction_recovery_multiple_databases.out /home/circleci/project/src/test/regress/results/multi_transaction_recovery_multiple_databases.out
--- /home/circleci/project/src/test/regress/expected/multi_transaction_recovery_multiple_databases.out.modified 2023-09-01 13:15:35.483441988 +0000
+++ /home/circleci/project/src/test/regress/results/multi_transaction_recovery_multiple_databases.out.modified  2023-09-01 13:15:35.495442069 +0000
@@ -330,22 +330,21 @@
 (2 rows)
 
 DROP DATABASE db1;
 SELECT pg_terminate_backend(pid)
 FROM pg_stat_activity
 WHERE pid <> pg_backend_pid()
   AND datname = 'db2' ;
  pg_terminate_backend 
 ----------------------
  t
- t
-(2 rows)
+(1 row)
 
 DROP DATABASE db2;
 \c - - - :worker_2_port
 SELECT pg_terminate_backend(pid)
 FROM pg_stat_activity
 WHERE pid <> pg_backend_pid()
   AND datname = 'db1' ;
  pg_terminate_backend 
 ----------------------
  t
```
maybe just hide/aggregate the query result, e.g. `SELECT count(pg_terminate_backend(pid)) > 0`

**Comment by ivyazmitinov** [comment]
@marcocitus, done, and I updated recover queries in the same fashion: there are some rouge transactions apart from the synthetic ones in `pg_dist_transaction`

File: src/test/regress/expected/multi_transaction_recovery.out [hunk 4]
```diff
@@ -41,9 +43,9 @@ CREATE TABLE should_be_sorted_into_middle (value int);
 PREPARE TRANSACTION 'citus_0_should_be_sorted_into_middle';
 SET citus.force_max_query_parallelization TO ON;
 -- Add "fake" pg_dist_transaction records and run recovery
-INSERT INTO pg_dist_transaction VALUES (1, 'citus_0_should_commit'),
+INSERT INTO pg_dist_transaction VALUES (:worker_group_id, 'citus_0_should_commit'),
                                        (0, 'citus_0_should_commit');
-INSERT INTO pg_dist_transaction VALUES (1, 'citus_0_should_be_forgotten'),
+INSERT INTO pg_dist_transaction VALUES (:worker_group_id, 'citus_0_should_be_forgotten'),
```
Comment (marcocitus) [inline src/test/regress/expected/multi_transaction_recovery.out:48]: the test file change is triggering our "flaky test detection", which enforces that all test files are repeatable and not-flaky by repeating the tests that are touched by the PR many times. We did not enforce repeatability on existing tests, but any change triggers enforcement.

if you prefer, you can revert this change to avoid triggering it (we'll deal with it later), or try to change the test file to be repeatable

Comment (ivyazmitinov) [reply src/test/regress/expected/multi_transaction_recovery.out]: @marcocitus, done

File: src/test/regress/expected/multi_transaction_recovery_multiple_databases.out [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+FROM pg_stat_activity
+WHERE pid <> pg_backend_pid()
+  AND datname = 'db1' ;
+ ?column?
+---------------------------------------------------------------------
+ t
+(1 row)
+
+DROP DATABASE db1;
+SELECT count(pg_terminate_backend(pid)) > 0
```
Comment (marcocitus) [inline src/test/regress/expected/multi_transaction_recovery_multiple_databases.out:355]: is there any risk of this being 0? maybe >= 0 if it's not important

(we sometimes run our tests with postgres running under valgrind, which creates all sorts of weird timings)

**Review by marcocitus — APPROVED** [review]

File: src/test/regress/expected/multi_transaction_recovery_multiple_databases.out (hunk 5 above)
Comment (ivyazmitinov) [reply src/test/regress/expected/multi_transaction_recovery_multiple_databases.out:355]: Good point: since it is a clean up, essentially, it shouldn't matter. Let me do a quick fix...

**Comment by ivyazmitinov** [comment]
@marcocitus 
My pleasure :slightly_smiling_face: 
And I am extremely glad to hear that multi-db support is planned. I, actually, was thinking of tackling the architecture of the maintenance daemon next, adapting it to process all databases in a cluster, as it is a major pain for us. I hope you won't mind a few more contributions :slightly_smiling_face:
